### DISCUSSION TITLE
How to remove the default global styles and variables added by tailwindcss

### DISCUSSION THREAD

**Discussion by neel1998** [tailwindlabs/tailwindcss#12826] [Help] [answered]
currently when i run tailwindcss the a lot of default global style classes are added at the top of the generated css file. Is there a way I can avoid these styles being added to output file?

```
*,
::before,
::after {
  box-sizing: border-box;
  /* 1 */
  border-width: 0;
  /* 2 */
  border-style: solid;
  /* 2 */
  border-color: currentColor;
  /* 2 */
}

::before,
::after {
  --tw-content: '';
}

/*
1. Use a consistent sensible line-height in all browsers.
2. Prevent adjustments of font size after orientation changes in iOS.
3. Use a more readable tab size.
4. Use the user's configured `sans` font-family by default.
5. Use the user's configured `sans` font-feature-settings by default.
6. Use the user's configured `sans` font-variation-settings by default.
7. Disable tap highlights on iOS
*/

html,
:host {
  line-height: 1.5;
  /* 1 */
  -webkit-text-size-adjust: 100%;
  /* 2 */
  -moz-tab-size: 4;
  /* 3 */
  -o-tab-size: 4;
     tab-size: 4;
  /* 3 */
  font-family: ui-sans-serif, system-ui, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
  /* 4 */
  font-feature-settings: normal;
  /* 5 */
  font-variation-settings: normal;
  /* 6 */
  -webkit-tap-highlight-color: transparent;
  /* 7 */
}

/*
1. Remove the margin in all browsers.
2. Inherit line-height from `html` so users can set them as a class directly on the `html` element.
*/

body {
  margin: 0;
  /* 1 */
  line-height: inherit;
  /* 2 */
}

/*
1. Add the correct height in Firefox.
2. Correct the inheritance of border color in Firefox. (https://bugzilla.mozilla.org/show_bug.cgi?id=190655)
3. Ensure horizontal rules are visible by default.
*/

hr {
  height: 0;
  /* 1 */
  color: inherit;
  /* 2 */
  border-top-width: 1px;
  /* 3 */
}

/*
Add the correct text decoration in Chrome, Edge, and Safari.
*/

abbr:where([title]) {
  -webkit-text-decoration: underline dotted;
          text-decoration: underline dotted;
}

/*
Remove the default font size and weight for headings.
*/

h1,
h2,
h3,
h4,
h5,
h6 {
  font-size: inherit;
  font-weight: inherit;
}
```

**Comment by wongjn** [comment]
Those rules can be removed by [disabling preflight](https://tailwindcss.com/docs/preflight#disabling-preflight).

**Reply by neel1998** [reply]
oh thanks a lot. Also is there a way I can generate all the classes based on my config without relying on the content? including classes with all the pseudo selectors and screen variants?

**Reply by wongjn** [reply]
That would not really be possible for v3 – there's an infinite amount of classes possible (due to the arbitrary features) and even without these, if you really wanted every single class possible, you would need to generate every possible combination of the variants, (like `dark:first:hover:focus:md:visited:last:only:…:m-1`) so there would be all the "base" classes plus all these permutations of variants and thus the total number of possible classes would be something like (71![^1] + 1) × number of "base" classes (maths probably not correct here, but something like this). The "real" number would be slightly less since some variants don't make sense together, i.e. `sm:md:lg:xl:m-1` but this obscenely large number should demonstrate the point.

[^1]: 71 factorial, equal to 850478588567862317521167644239926010288584608120796235886430763388588680378079017697280000000000000000

**Reply by neel1998** [reply]
no that's why I was asking if there was a way to generate ONLY those classes which are defined in my custom config? not all the base classes provided by tailwind? I tried setting `presets: []` and then provided empty object for all the keys of the default config provided here https://github.com/tailwindlabs/tailwindcss/blob/master/stubs/config.full.js . And just kept borderRadius as is for now. The generated css does generate all the required classes for `rounded-` and none for the rest of the props. But it also adds classes for display and some other props which are not there in config ... so my questions are,

1. where are these display and other rest of the classes defined which are not included in the default config and yet reflect in the generated css

2. How do I generate the classes for pseudo selectors also for all the classes i have in the config. In the current example, I would want `hove:rounded-` and other similar classes also generated.

**Reply by wongjn** [reply, marked as answer]
1. The display and other "static" classes are generated from "core plugins". [You can disable unneeded ones](https://tailwindcss.com/docs/theme#disabling-an-entire-core-plugin).
2. You can add variants to regular expression safelists [via the `variants` key](https://tailwindcss.com/docs/content-configuration#using-regular-expressions).

**Reply by neel1998** [reply]
oh awesome. Thanks a lot! this works
